src/app-layer-ssl.c

```diff
@@ -255,6 +259,249 @@ static void SSLSetTxDetectFlags(void *vtx, uint8_t dir, uint64_t flags)
+                                                         sizeof(char));
+    if (ssl_state->server_connp.cert0_fingerprint == NULL)
+        return -1;
+
+    uint8_t *hash = ComputeSHA1(input, cert_len);
+    if (hash == NULL)
+        return 0;
+
+    int i, x;
+    for (i = 0, x = 0; x < SHA1_LENGTH; x++)
+    {
+        i += snprintf(ssl_state->server_connp.cert0_fingerprint + i,
+                      SHA1_STRING_LENGTH - i, i == 0 ? "%02x" : ":%02x",
+                      *(hash + x));
+    }
+
+    SCFree(hash);
+
+    return 0;
+}
+
+static inline int TlsDecodeHSCertificateAddCertToChain(SSLState *ssl_state,
+                                                       uint8_t *input,
+                                                       uint32_t cert_len)
+{
+    SSLCertsChain *cert = SCCalloc(1, sizeof(SSLCertsChain));
+    if (cert == NULL)
+        return -1;
+
+    cert->cert_data = input;
+    cert->cert_len = cert_len;
+    TAILQ_INSERT_TAIL(&ssl_state->server_connp.certs, cert, next);
+
+    return 0;
+}
+
+static int TlsDecodeHSCertificate(SSLState *ssl_state,
+                                  const uint8_t * const initial_input,
+                                  const uint32_t input_len)
+{
+    uint8_t *input = (uint8_t *)initial_input;
+
+    Asn1Generic *cert;
+
+    if (!(HAS_SPACE(3)))
+        return 1;
+
+    uint32_t cert_chain_len = *input << 16 | *(input + 1) << 8 | *(input + 2);
+    input += 3;
+
+    if (!(HAS_SPACE(cert_chain_len)))
+        return 0;
+
+    uint32_t processed_len = 0;
+    while (processed_len < cert_chain_len)
+    {
+        if (!(HAS_SPACE(3)))
+            goto invalid_cert;
+
+        uint32_t cert_len = *input << 16 | *(input + 1) << 8 | *(input + 2);
+        input += 3;
+
+        if (!(HAS_SPACE(cert_len)))
+            goto invalid_cert;
+
+        uint32_t err = 0;
+        int rc = 0;
+
+        /* only store fields from the first certificate in the chain */
+        if (processed_len == 0) {
+            cert = DecodeDer(input, cert_len, &err);
+            if (cert == NULL) {
+                TlsDecodeHSCertificateErrSetEvent(ssl_state, err);
+                goto next;
+            }
+
+            rc = TlsDecodeHSCertificateSubject(ssl_state, cert);
+            if (rc != 0)
+                goto error;
+
+            rc = TlsDecodeHSCertificateIssuer(ssl_state, cert);
+            if (rc != 0)
+                goto error;
+
+            rc = TlsDecodeHSCertificateSerial(ssl_state, cert);
+            if (rc != 0)
+                goto error;
+
+            rc = TlsDecodeHSCertificateValidity(ssl_state, cert);
+            if (rc != 0)
+                goto error;
+
+            rc = TlsDecodeHSCertificateFingerprint(ssl_state, input, cert_len);
+            if (rc != 0)
+                goto error;
+
+            DerFree(cert);
+        }
+
+        rc = TlsDecodeHSCertificateAddCertToChain(ssl_state, input, cert_len);
+        if (rc != 0)
+            return -1;
```

Comment:
why doesn't this jump to error?

Reply:
Because it has nothing to clean up. I'm changing it to jump to error, since that makes it easier to understand what's happening.